Backend for species specific iconbases

InfinityArch opened the pull request:
this change makes it possible to define species specific iconbases (.dmi filepaths). Alternate iconbase file paths must be added into an ordered list (alternate_iconbases) in the datum/species. For now this only works with clothing.

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

These minor change makes it possible to define species specific iconbases. Alternate iconbases must be added into an ordered list in the species datum. For now this only works with clothing, since the only check for alternate iconbases I added was to build_worn_icon()

an example that can be used to validate this PR is to add 

alternate_iconbases = list('icons/mob/clothing/under/captain.dmi' = 'icons/mob/clothing/head')

to a species datum, which will result in the captain's hat being displayed on the icon when wearing the captain's uniform as the race you chose for this test.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Currently /tg/'s .dmi files are kind of cluttered, to put things mildly. There's been some progress in that regards with the recent breakup of the uniform sprites, and this will make it possible to do for species bodyparts, as well as basically anything else that's being applied to human type mobs. It also makes it possible to have a species with a body shape that doesn't fit into the normal clothing boundaries (ie Vox from the paradise codebaes), if the creator is willing to put the work in to make all the necessary conversions for every state on the sheet, and the huge number of extra icon states involved won't end up cluttering the main ones.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Added the ability to give species their own iconbases
add: added the variable list/alternate_iconbases ( initial state of list() ) to datum/species where you associate current iconbases with their alternate versions. 
add: Added about 5 lines of code into build_worn_icon() to check for alternate iconbases, plus a bunch of documentation
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

InfinityArch committed:
adds the ability to give specific species their own spritesheets

this change makes it possible to define species specific spritesheets. Alternate spritesheets must be added into an ordered list in the species datum. For now this only works with clothing.

carlarctg commented:
finally i can be a snowflake lizard and have my tail show up in hardsuits

InfinityArch commented:
> finally i can be a snowflake lizard and have my tail show up in hardsuits

Sure. If you want to put in the work, you can do that, and you can keep your speical spritesheets in their own folder where they won't add clutter. Compare this (4 lines of code not including comments) to digitigrade where it's a massive, convoluted mess.

Qustinnus commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -15,6 +15,9 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	var/hair_color	// this allows races to have specific hair colors... if null, it uses the H's hair/facial hair colors. if "mutcolor", it uses the H's mutant_color
 	var/hair_alpha = 255	// the alpha used by the hair. 255 is completely solid, 0 is transparent.
 
+	var/list/alternate_spritesheets = list() // this is an association list, put in a spritesheet path matched with an alternate sheet to swap out the spritesheet for this species for an alternative one
```
autodoc

Qustinnus commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -514,6 +514,11 @@ generate/load female uniform sprites matching all previously decided variables
 	if(!file2use)
 		file2use = default_icon_file
 
+	//Check for an alternate spritesheet path in the species datum
```
ditto

InfinityArch commented:
My apologies for wasting your time. As it turns out, there is an oversight in this system; it requires an icon file to be defined previously in the system at some point, I need to make some additional changes to make this work.

AnturK commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -514,6 +514,11 @@ generate/load female uniform sprites matching all previously decided variables
 		file2use = default_icon_file
 
+	//Check for an alternate spritesheet path in the species datum
+	var/mob/living/carbon/human/H = loc
```
Will runtime for non humans.

AnturK commented:
Also i'd rather you didn't use the word spritesheet that has more specific meaning than just dmi file in code.

InfinityArch committed:
Changed based on reviewer feedback

Following feedback from reviewers, I am improving the documentation of the alternate sprite sheet system, changed the name to alternate iconbases to reflect the fact that this system works with icon spritesheets rather than any spritesheet, and moved the steps for checking for alternate iconbases into its own proc.

InfinityArch committed:
Changed update_icons to reflect reviewers comments

improved documentation of this system, added a proc, and added a check to avoid runtimes from nonhumans

AnturK commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -514,6 +514,11 @@ generate/load female uniform sprites matching all previously decided variables
+	//Check for an alternate spritesheet path in the species datum
+    var/mob/living/carbon/human/H = loc
+    if(istype(H))
+	    file2use = get_species_iconbase(file2use, isinhands)
```
This is performance baggage enough without wrapping it in a proc.

AnturK reviewed (changes requested):
Also bad indents.

InfinityArch committed:
removed extra proc as per reviewer

Removed the get_species_iconbase() proc, as per the reviewer's request, removed a bad indentation, and reworked the check to hopefully be more efficient than my original version and the version that used the proc.

InfinityArch committed:
updated documentation

build_worn_icon() is now the proc that makes use of alternate_iconbases instead of the check being wrapped inside of a list.

AnturK commented:
After some consideration i'll be closing this. I gave you some explanation on discord before but to reiterate for everyone else. If you want to add support for species specific variants it needs to :
a) have minimal impact on the normal icon building.
b) cause no duplication of dmi's / allow for selection what icon states are replaced.
